File: app/src/main/java/com/homeworkshop/notemvvm/NoteAdapter.java
package com.homeworkshop.notemvvm;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.AsyncDifferConfig;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class NoteAdapter extends ListAdapter<Note, NoteAdapter.NoteHolder> {
    private OnItemClickListener listener;

    public NoteAdapter() {
        super(DIFF_CALLBACK);
    }

    // obiekt ktory porównuje elementy z nowej listy z elementami ze starej listy
    // przekazujemy go do superklass w konstruktorze
    private static final DiffUtil.ItemCallback<Note> DIFF_CALLBACK = new DiffUtil.ItemCallback<Note>() {
        //metoda zwraca true jeżeli elementy są takie same. Nie znaczy że zawartość jest taka sama ale że to są te same obiekty w naszej liscie
        // dlatego porównujemy je po id. Wiemy że wtedy to jest to samo Entry z listy , nawet z różna zawartościa
        @Override
        public boolean areItemsTheSame(@NonNull Note oldItem, @NonNull Note newItem) {
            return oldItem.getId() == newItem.getId();
        }

        //zwraca true jeżeli nic się nie zmieniło w zawartości elementu
        @Override
        public boolean areContentsTheSame(@NonNull Note oldItem, @NonNull Note newItem) {
            return oldItem.getTitle().equals(newItem.getTitle()) && oldItem.getDescription().equals(newItem.getDescription()) && oldItem.getPriority() == newItem.getPriority();
        }
    };

    @NonNull
    @Override
    public NoteHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View item_view = LayoutInflater.from(parent.getContext()).inflate(R.layout.note_item, parent, false);
        return new NoteHolder(item_view);
    }

    @Override
    public void onBindViewHolder(@NonNull NoteHolder holder, int position) {
        Note currentNote = getItem(position);
        holder.textViewTitle.setText(currentNote.getTitle());
        holder.textViewDescription.setText(currentNote.getDescription());
        holder.textViewPriority.setText(String.valueOf(currentNote.getPriority()));
    }

    //zwraca notatkę na danej pozycji - potrzebne np do usuwania danej notatki
    public Note getNoteAt(int position) {
        return getItem(position);
    }

    class NoteHolder extends RecyclerView.ViewHolder {
        private TextView textViewTitle;
        private TextView textViewDescription;
        private TextView textViewPriority;

        public NoteHolder(@NonNull View itemView) {
            super(itemView);
            textViewTitle = itemView.findViewById(R.id.text_view_title);
            textViewDescription = itemView.findViewById(R.id.text_view_description);
            textViewPriority = itemView.findViewById(R.id.text_view_priority);
            //ustawiamy zdarzenie onClick dla całej ramki notatki
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    int position = getAdapterPosition();
                    if (listener != null && position != RecyclerView.NO_POSITION) {
                        listener.onItemClick(getItem(position));
                    }
                }
            });
        }
    }

    //Aby mieć zdarzenie onClick na całej notatce potrzebujemy interface
    public interface OnItemClickListener {
        void onItemClick(Note note);
    }

    //metoda ustawiajaca referencje do OnItemClickListenera
    public void setOnItemClickListener(OnItemClickListener listener) {
        this.listener = listener;
    }

}

File: app/src/main/java/com/homeworkshop/notemvvm/NoteAdapter.kt
package com.homeworkshop.notemvvm

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.homeworkshop.notemvvm.NoteAdapter.NoteHolder

class NoteAdapter : ListAdapter<Note, NoteHolder>(DIFF_CALLBACK) {
    private var listener: OnItemClickListener? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NoteHolder {
        val item_view = LayoutInflater.from(parent.context).inflate(R.layout.note_item, parent, false)
        return NoteHolder(item_view)
    }

    override fun onBindViewHolder(holder: NoteHolder, position: Int) {
        val currentNote = getItem(position)
        holder.textViewTitle.text = currentNote!!.title
        holder.textViewDescription.text = currentNote.description
        holder.textViewPriority.text = currentNote.priority.toString()
    }

    //zwraca notatkę na danej pozycji - potrzebne np do usuwania danej notatki
    fun getNoteAt(position: Int): Note {
        return getItem(position)
    }

    inner class NoteHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val textViewTitle: TextView
        val textViewDescription: TextView
        val textViewPriority: TextView

        init {
            textViewTitle = itemView.findViewById(R.id.text_view_title)
            textViewDescription = itemView.findViewById(R.id.text_view_description)
            textViewPriority = itemView.findViewById(R.id.text_view_priority)
            //ustawiamy zdarzenie onClick dla całej ramki notatki
            itemView.setOnClickListener {
                val position = adapterPosition
                if (listener != null && position != RecyclerView.NO_POSITION) {
                    listener!!.onItemClick(getItem(position))
                }
            }
        }
    }

    //Aby mieć zdarzenie onClick na całej notatce potrzebujemy interface
    interface OnItemClickListener {
        fun onItemClick(note: Note)
    }

    //metoda ustawiajaca referencje do OnItemClickListenera
    fun setOnItemClickListener(listener: OnItemClickListener) {
        this.listener = listener
    }

    companion object {
        // obiekt ktory porównuje elementy z nowej listy z elementami ze starej listy
        // przekazujemy go do superklass w konstruktorze
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<Note> = object : DiffUtil.ItemCallback<Note>() {
            //metoda zwraca true jeżeli elementy są takie same. Nie znaczy że zawartość jest taka sama ale że to są te same obiekty w naszej liscie
            // dlatego porównujemy je po id. Wiemy że wtedy to jest to samo Entry z listy , nawet z różna zawartościa
            override fun areItemsTheSame(oldItem: Note, newItem: Note): Boolean {
                return oldItem.id == newItem.id
            }

            //zwraca true jeżeli nic się nie zmieniło w zawartości elementu
            override fun areContentsTheSame(oldItem: Note, newItem: Note): Boolean {
                return oldItem.title == newItem.title && oldItem.description == newItem.description && oldItem.priority == newItem.priority
            }


        }
    }
}